### ISSUE TITLE
X25519 EVP

### ISSUE THREAD

**Issue by aep** [sfackler/rust-openssl#947]
X25519 is a bit stitched together in openssl. 

I believe you cannot use EC_GROUP_new_by_curve_name.

Any idea how to implement Pkey for X25519 properly? I'm going to toy around a bit

**Comment by sfackler** [comment]
I'm not personally familiar, but if you track down some C code that does what you want that'd be a good start on figuring out what APIs need to be exposed.

**Comment by hahseba** [comment]
Did you make any progress on this @aep? I was also bitten by this today, it seems I cannot even get the raw public key from a certificate.

**Comment by aep** [comment]
No, i'm no longer using TLS at all. Instead i went with NOISE which has a pure rust implementation (snow)

**Comment by LinusU** [comment]
@sfackler I would like to use the Curve25519 functions in an app I'm porting from a C codebase. It was previously using another C-library, but I've just managed to port it over to use OpenSSL instead (and with that the last piece of C was removed and my app is not 100% Rust 😍).

I've currently implemented it quite ugly and just defined an `extern "C"` block with what OpenSSL functions I needed and used them straight from Rust 😄 

This file should be a good idea on what I need:

https://github.com/LinusU/rust-raop-player/blob/b65e0bd3a3cebcc9d009eec7931a371a9c2e4d6e/src/curve25519.rs

It's not super nicely organized at the moment, and does a number of extra copies and generally handing out raw bytes instead of nice structures. I just wanted it to mimic the previous library at the moment.

I'd be very happy to work with you to upstream this and get support into the OpenSSL crate, what do you think? ☺️
